sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -61,6 +62,11 @@ private[sql] class SharedState(val sparkContext: SparkContext) extends Logging {
       // When neither spark.sql.warehouse.dir nor hive.metastore.warehouse.dir is set,
       // we will set hive.metastore.warehouse.dir to the default value of spark.sql.warehouse.dir.
       val sparkWarehouseDir = sparkContext.conf.get(WAREHOUSE_PATH)
+      logInfo(s"Setting hive.metastore.warehouse.dir ($hiveWarehouseDir) to the value of " +
+        s"${WAREHOUSE_PATH.key} ('$sparkWarehouseDir').")
+      if (hiveWarehouseDir != null) {
+        sparkContext.hadoopConfiguration.set("hive.metastore.warehouse.dir", sparkWarehouseDir)
+      }
       sparkContext.conf.set("hive.metastore.warehouse.dir", sparkWarehouseDir)
```

Comment:
I look into the code, and make some tests.
The `hive.metastore.warehouse.dir` in sparkConf still take effect in Spark, it is not useless.
The reason is that:
1. when we run spark with HiveEnabled, it will create `ShareState`
2. when create `ShareState`, it will create a `HiveExternalCatalog` 
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala#L85
3.when create `HiveExternalCatalog`, it will Create `HiveClientImpl` by `HiveUtils`
https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L65
4.when create `HiveClientImpl`, it will call [SessionState.start(state)](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L193)
5. and then in the `SessionState.start(state)`, it will `create a default database` using `hive.metastore.warehouse.dir` in `hiveConf` which is created in `HiveClientImpl` https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L189
6. while the `hiveConf` created in `HiveClientImpl`  from `hadoopConf` and `sparkConf `, and `sparkConf` will overwrite the value of the same key in hadoopConf. So it means that it actually will use `hive.metastore.warehouse.dir` in `sparkConf` to create the default database, if we does not overwrite the value in `sparkConf` in [SharedState(https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala#L64), the database location is not we expected which is the warehouse path. So here `sparkContext.conf.set("hive.metastore.warehouse.dir", sparkWarehouseDir)` should be retained

we can also find that,the default database does not created in `SharedState`, here condition is false, will not hit the create database logic.
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala#L96